# MC316 Final Exam – Part B: PYTHON

* You will have 1 hour and 40 minutes to finish the test.
* This part includes 10 questions. Each question is worth 15 points.
* You get extra 30 points for full answers.
* Please read the instructions carefully. 

## I. SET UP


### Import pandas library for use in scripting. 

In [3]:
import pandas as pd

### Import the datasets and name them.
For this test, you will analyze revenue and expenditure data released by Kansas Department of Administration for the fiscal year 2018. While the original data can be directly downloaded [here](http://kanview.ks.gov/DataDownload.aspx), for your convenience, the instructor had them uploaded to a server at the following urls. 

#### For Kansas State University:
1. Expenditure data: https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_367_2018.csv
2. Revenue data: https://datajournalism.tech/wp-content/uploads/2019/05/RevData_367_2018.csv

#### For University of Kansas:
3. Expenditure data: https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_682_2018.csv
4. Revenue data: https://datajournalism.tech/wp-content/uploads/2019/05/RevData_682_2018.csv

You will need to import all 4 datasets and name them in an easy way to call out later.

*Please refer to the [Glossary](http://kanview.ks.gov/Glossary.aspx#f) for data description.

In [19]:
KSUEXP = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_367_2018.csv")

In [20]:
KSUREV = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/RevData_367_2018.csv")

In [21]:
KUEXP = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_682_2018.csv")

In [22]:
KUREV = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/RevData_682_2018.csv")

## II. REQUIRED QUESTIONS
* Make sure you type the answers in the cells right below the codes.
* If you need to add extra cells, click the + button in the toolbar up the top. 

### Q1: What is the total revenue recorded for Kansas State University in the fiscal year of 2018? How does it compare to the total revenue recorded for University of Kansas in the same year?

In [23]:
KSUREV.Total.sum()

854324643.84

In [24]:
KUREV.Total.sum()

808767902.55

In [25]:
854324643.84-808767902.55

45556741.29000008

The sum of the total revenue for Kansas State University in 2018 was 854,432,642.84. The sum of total revenue for Uinversity of Kansas for this year was 808,767,902.55. The difference in revenue shows K-State had 45,556,741.29 more revenue in 2018.

### Q2: What is the total expense recorded for Kansas State University in the fiscal year of 2018? How does it compare to the total expense recorded for University of Kansas in the same year?

In [26]:
KSUEXP.Total.sum()

1018584373.02

In [27]:
KUEXP.Total.sum()

990238875.6099999

In [28]:
1018584373.02-990238875.6099999

28345497.410000086

<b>Answer:</b> The total expenses in the fiscal year 2018 for Kansas State University was 1,018,584,373.02. The total expenses for University of Kansas in this year was 990,238.875.60. This shows that the difference between the two school's expenses is 28,345,497.41, and Kansas State had more expenses. 

### Q3: Where did the two universities' primary revenues come from? 
*Refer to the description [here](http://kanview.ks.gov/Revenues/AcctCodeRevenues.aspx) to write your answer.

In [41]:
KSUREV.Primary_Account_Code_Descr.value_counts()

AGENCY EARNINGS                   637
REV FROM USE OF MONEY & PROP      354
REVENUE TRANSFERS                 341
NON REVENUE RECEIPTS              259
OTHER REVENUE                     160
GIFT DONATION FEDERAL GRANTS      152
TAXES                               7
Name: Primary_Account_Code_Descr, dtype: int64

In [42]:
KUREV.Primary_Account_Code_Descr.value_counts()

AGENCY EARNINGS                   376
REV FROM USE OF MONEY & PROP      251
NON REVENUE RECEIPTS              157
REVENUE TRANSFERS                 153
OTHER REVENUE                      57
TAXES                              45
GIFT DONATION FEDERAL GRANTS       38
Name: Primary_Account_Code_Descr, dtype: int64

<b>Answer:</b> The primary source of revenue for both universities was Agency Earnings. 

### Q4: What was the biggest source of revenues for Kansas State University? Compare to the one of University of Kansas.

Remove hashtag and run the following code to suppress scientific notation:

In [49]:
pd.set_option('display.float_format', lambda x: '%.3f' % x)

In [59]:
KSUREV.groupby("Primary_Account_Code_Descr").Total.sum()

Primary_Account_Code_Descr
AGENCY EARNINGS                  412565112.820
GIFT DONATION FEDERAL GRANTS     278778848.030
NON REVENUE RECEIPTS              71947564.050
OTHER REVENUE                      8432030.160
REV FROM USE OF MONEY & PROP      34139318.040
REVENUE TRANSFERS                 48458616.490
TAXES                                 3154.250
Name: Total, dtype: float64

In [60]:
KUREV.groupby("Primary_Account_Code_Descr").Total.sum()

Primary_Account_Code_Descr
AGENCY EARNINGS                  408409337.320
GIFT DONATION FEDERAL GRANTS     193836668.320
NON REVENUE RECEIPTS              69753247.670
OTHER REVENUE                     68241930.500
REV FROM USE OF MONEY & PROP      40950779.540
REVENUE TRANSFERS                 26013855.400
TAXES                              1562083.800
Name: Total, dtype: float64

In [61]:
412565112.820-408409337.320

4155775.5

<b>Answer:</b> The biggest source of revenue for Kansas State University was Agency Earning with 412,565,112.82. The largest for KU was also Agency earnings at 408,409,337.32. K-State had a higher number by 4,155,775.50. 

### Q5: What accounts contributed to Kansas State University's biggest source of revenues?
If you run into a situation where a filter function returns 0 results, you may want to check if there are leading and trailing spaces in the column. To clean the leading and trailing spaces, you will need to remove the hashtag, revise the names, and run the following code:

In [63]:
#yourdataframe["yourcolumnname"] = yourdataframe["yourcolumnname"].str.strip()

After cleaning the spaces, you can start filtering your datasets to find the answer for Q5.

<b>Answer:</b> Agency Earnings, Gift donation, Non Revenue Receipts, Other Revenue. 

### Q6: Where did the money go? (Answer for both universities)
*Refer to the description [here] (http://kanview.ks.gov/Expenditures/AcctCodeExpenditures.aspx) to write your answer.

In [69]:
KSUEXP.groupby("Detail_Account_Code_Descr").Total.sum()

Detail_Account_Code_Descr
ACCOUNTANTS AND AUDITORS              9052.000
ADVERTISING                        2129373.840
ADVERTISING AND MARKETING            11953.000
AG EQUIP AND MACHINERY             1019999.810
AG EQUIP AND MACHINERY NON CAP      146858.820
AG SUPPLIES AND MATERIALS          1731336.670
AIR RAIL & BUS FARE INTERNATL      1085212.280
AIR RAIL & BUS FARE OUT STATE      1717919.260
AIR RAIL AND BUS FARE IN STATE       28618.820
AIRCRAFT FUEL                       298995.650
ANIMAL DRUGS                        655544.990
ARCHITECTS AND ENGINEERS           3013111.230
ASSESSMENTS FOR IMPROVEMENTS          1778.080
ATTORNEYS AND LAWYERS               356096.490
BANK FEES                              292.700
BLDGS AND GROUNDS MATERIALS        1605191.090
BOND INTEREST                     15417585.700
BOND PRINCIPAL                    15960000.000
BOOKS & LIBRARY MAT NON CAP         223804.920
BUILDING SPACE RENTAL              1261138.430
BUILDINGS                         

In [66]:
KUEXP.groupby("Detail_Account_Code_Descr").Total.sum()

Detail_Account_Code_Descr
ACCOUNTANTS AND AUDITORS            104594.090
ADVERTISING                        1849687.370
ADVERTISING AND MARKETING                0.000
AG EQUIP AND MACHINERY               23702.400
AG EQUIP AND MACHINERY NON CAP        9614.050
AG SUPPLIES AND MATERIALS            91514.860
AIR RAIL & BUS FARE INTERNATL       667468.030
AIR RAIL & BUS FARE OUT STATE      1034460.070
AIR RAIL AND BUS FARE IN STATE       33000.370
AIRCRAFT FUEL                       246502.750
ANIMAL DRUGS                            83.470
ANTIFREEZE                            1289.670
ARCHITECTS AND ENGINEERS           4004007.880
ASSESSMENTS FOR IMPROVEMENTS             0.000
ASSIGNED COUNSEL-CNTRCT LAWYER        1062.500
ASSIGNED COUNSEL-NONCNTRCT LAW        1764.000
ATTORNEYS AND LAWYERS                  117.330
BIO-DIESEL                            5081.830
BLDGS AND GROUNDS MATERIALS        4865591.890
BOND INTEREST                     13514706.920
BOND PRINCIPAL                    

<b>Answer:</b> The money was spent on a high variety of things. There are lots of different categories for the expenses. Some things the momey was spent on was Trucks, Water, Workers Compensation, Accounting, and Advertising. The main two things money was spent on were Unclassified Regular and Other loans.

### Q7: What costs two universities the most? 

<b>Answer:</b> The unclassified regular category costs the two universities the most. 

### Q8: Calculate the percentage of this cost over each university's total expense.

In [71]:
305121246.380/990238875.6099999

0.3081289312056562

In [72]:
305142755.620/1018584373.02

0.299575335831319

<b>Answer:</b> This cost makes up about 30% of the total expenses for KU. For Kansas State it makes up just under 30% (29.95%) of the total expenses. 

### Q9: Breakdown the cost you found in Q7 for both universities.
Again, if you run into a situation where a filter function returns 0 results, you may want to check if there are leading and trailing spaces in the column. To clean the leading and trailing spaces, you will need to remove the hashtag, revise the names, and run the following code:

In [3]:
#yourdataframe["yourcolumnname"] = yourdataframe["yourcolumnname"].str.strip()

<b>Answer:</b>

### Q10: Compare the breakdown costs from Q9 side by side.
When you merge the two subsets, make sure to add the attribute <b>how ="outer"</b> so that you can keep every item from both sets.
*For example: merge = pd.merge(df1, df2, how="outer", on="columnname")

<b>Answer:</b>

## III. BONUS QUESTIONS
* Each question's worth 15 points.


### B1: Which university made a profit in the fiscal year of 2018?

In [43]:
KSUREV.Total.sum()

854324643.84

In [44]:
KSUEXP.Total.sum()

1018584373.02

In [45]:
854324643.84-1018584373.02

-164259729.17999995

In [46]:
KUREV.Total.sum()

808767902.55

In [47]:
KUEXP.Total.sum()

990238875.6099999

In [48]:
808767902.55-990238875.6099999

-181470973.05999994

<b>Answer:</b> Neither university made a profit in 2018.

### B2: Which software program do you like the most in this course? Why?

<b>Answer:</b> I think I prefered Tableau over any of the other softwares we used. I liked it because it was one of the more simple softwares we used and seemed less confusing to me. Also, you can make some really neat graphics with Tableau and I feel like I mastered this one the best and have the most abilities in this software. 

### END OF THE TEST